src/OpenTelemetry.Instrumentation.Wcf/TelemetryPropagationReader.cs

```diff
@@ -0,0 +1,108 @@
+// <copyright file="TelemetryPropagationReader.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.ServiceModel.Channels;
+using OpenTelemetry.Instrumentation.Wcf.Implementation;
+using OpenTelemetry.Internal;
+
+namespace OpenTelemetry.Instrumentation.Wcf;
+
+/// <summary>
+/// Pre-defined PropagationReader callbacks.
+/// </summary>
+public static class TelemetryPropagationReader
+{
+    private static readonly Func<Message, string, IEnumerable<string>> DefaultReader = Compose(HttpRequestHeaders, SoapMessageHeaders);
+
+    /// <summary>
+    /// Reads the values from the SOAP message headers. If the message is not a SOAP message it returns null.
+    /// </summary>
+    /// <param name="request">The incoming <see cref="Message"/> to read trace context from.</param>
+    /// <param name="name">The header name being requested.</param>
+    /// <returns>An enumerable of all the values for the requested header, or null if the requested header was not present on the request.</returns>
+    public static IEnumerable<string> SoapMessageHeaders(Message request, string name)
```

Comment:
Conceptually why is this returning an IEnumerable when both the HTTP and SOAP Header mechanism only allow for a single instance item?

Reply:
It only returns IEnumerable because the `TextMapPropagator.Extract()` method expects IEnumerable (see https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/Context/Propagation/TextMapPropagator.cs#L50). I don't really understand that use-case but I figured it must be like that on the propagators for a reason, so I would just maintain that capability.

Reply:
Ok, today I learned - the reason is baggage. The value for baggage is unbounded and can overflow a single entry, so this interface gives opentelemetry the ability to have full control over how to split the value across multiple entries and subsequently reconstitute it on the receiving side.  It will call the writer multiple times with the same header name - once for each part of the split-up value....and then on the receiving side it will just call the reader once to retrieve all the parts from the carrier at once.  This way the readers+writers don't need to implement any of the splitting logic, they just need to support reading and writing multiple values for the same header name.  So, I guess, the interface is correct but the implementation will not support large baggage values. This new implementation, however, has the same baggage support as the existing implementation in `main` so I'd rather not tackle "proper baggage support" as part of this PR, I'd rather do it in a separate PR.